Update openebs plugin

vharsh opened the pull request:
Changes:
- Bumped up release v0.2.0 -> v0.3.1
- Bundle project LICENSE
- Use auth plugins in all client initialization
- Flattened archive contents

Signed-off-by: Harsh Vardhan <harsh.vardhan@mayadata.io>

<!--

PLUGIN DEVELOPERS: If you are submitting a new plugin

- Make sure you read the Plugin Naming Guide: https://krew.sigs.k8s.io/docs/developer-guide/develop/naming-guide/
- Verify you can install your plugin locally: kubectl krew install --manifest=[...] --archive=[...]

-->

ahmetb commented:
:robot: _Beep beep! I’m a robot speaking on behalf of @ahmetb._ :robot:

-----

This pull request **does not** seem to be a straightforward version bump. I'll have a human review this.

_Why wasn't this detected as a plugin version bump:_

>file `plugins/openebs.yaml` is not a straightforward version bump: change on line `5` does not seem like a version bump: [`-    bin: kubectl-openebs_v0.2.0_Darwin_x86_64/kubectl-openebs`]

vharsh commented:
It did work with my [custom krew index](https://github.com/vharsh/krew) with the same `openebs.yaml` on `Linux + x86_64`.

```zsh
# Adding the custom index
➜  plugins git:(main) ✗ kubectl krew index add custom https://github.com/vharsh/krew.git
WARNING: You have added a new index from "https://github.com/vharsh/krew.git"
The plugins in this index are not audited for security by the Krew maintainers.
Install them at your own risk.

# Installing
➜  plugins git:(main) ✗ kubectl krew install custom/openebs  
Updated the local copy of plugin index "custom".
Updated the local copy of plugin index.
Installing plugin: openebs
Installed plugin: openebs
\
 | Use this plugin:
 | 	kubectl openebs
 | Documentation:
 | 	https://github.com/openebs/openebsctl
 | Caveats:
 | \
 |  | * This plugin works with newer OpenEBS releases, where volumes are
 |  |   provisioned by CSI.
 |  | * For resources that are not in default namespace, you must specify
 |  |   -n/--namespace explicitly (the current namespace setting is not
 |  |   yet used).
 | /
/

➜  plugins git:(main) ✗ cd ~/.krew/store/openebs/v0.3.0 
➜  v0.3.0  ls
kubectl-openebs  LICENSE  README.md
```

ahmetb commented:
:robot: _Beep beep! I’m a robot speaking on behalf of @ahmetb._ :robot:

-----

This pull request **does not** seem to be a straightforward version bump. I'll have a human review this.

_Why wasn't this detected as a plugin version bump:_

>file `plugins/openebs.yaml` is not a straightforward version bump: change on line `5` does not seem like a version bump: [`-    bin: kubectl-openebs_v0.2.0_Darwin_x86_64/kubectl-openebs`]

vharsh commented:
@ahmetb Can you point out what I'm missing this time? I don't have a Windows setup to test out relative paths, I cannot figure out why the builds for Windows are failing on a nix platform, I'm assuming I might have to change `/` with `\` for windows machines.
CI has these errors
```logs
	I0813 17:18:43.374440    1997 verifier.go:51] Compare sha256 (6c713ca5dbf9c2c6e554159be32cb101dba5d09ef20f25aaba65babd180ae06a) signed version
	I0813 17:18:43.374462    1997 downloader.go:203] detected "application/zip" file type
	I0813 17:18:43.374474    1997 downloader.go:54] Extracting zip archive to "/tmp/krew-downloads360838064"
	I0813 17:18:43.374639    1997 install.go:101] Deleting the download staging directory /tmp/krew-downloads360838064
	W0813 17:18:43.374743    1997 install.go:164] failed to install plugin "openebs": install failed: failed to unpack into staging dir: failed to unpack the plugin archive: failed to extract file: can't create file in zip destination dir: open /tmp/krew-downloads360838064/kubectl-openebs_v0.3.0_Windows_x86_64/LICENSE: no such file or directory
	F0813 17:18:43.374789    1997 root.go:77] open /tmp/krew-downloads360838064/kubectl-openebs_v0.3.0_Windows_x86_64/LICENSE: no such file or directory
```
And the zip bundle for windows does contain the `LICENSE` file.

```bash
➜  /tmp mktemp -d 
/tmp/tmp.ldYWhzwLDr
➜  /tmp cd tmp.ldYWhzwLDr 
➜  tmp.ldYWhzwLDr mv ../kubectl-openebs_v0.3.0_Windows_x86_64.zip .
➜  tmp.ldYWhzwLDr ls
kubectl-openebs_v0.3.0_Windows_x86_64.zip
➜  tmp.ldYWhzwLDr unzip kubectl-openebs_v0.3.0_Windows_x86_64.zip
Archive:  kubectl-openebs_v0.3.0_Windows_x86_64.zip
  inflating: kubectl-openebs_v0.3.0_Windows_x86_64/LICENSE  
  inflating: kubectl-openebs_v0.3.0_Windows_x86_64/README.md  
  inflating: kubectl-openebs_v0.3.0_Windows_x86_64/kubectl-openebs.exe  
➜  tmp.ldYWhzwLDr pwd
/tmp/tmp.ldYWhzwLDr
```

ahmetb commented:
> I'm assuming I might have to change `/` with `\` for windows machines.

That's probably not the reason. That said, I _also_ don't see where the problem is. Let me try working on this locally. Might be a bug somehow.

Also, if there's a way for you to zip/tar up this release without creating a directory inside the archive, it would be much better. (If you omit the `files:` field, we extract the archive as is.)

ahmetb commented:
OK, this is a bug in our zip extractor. It seems that we have not handled for the case where we get a ZIP file with a non-flat directory structure in it. (Surprising to me as well. I'll open a bug in krew repo.)

I am assuming this bug happened because the archive/zip package gives you entries as *os.File, but the ZIP archives don't have directory entries unlike TAR archives. (I think we didn't know that at the time.) We call f.FileInfo().IsDir() but it would practically never be used. https://github.com/kubernetes-sigs/krew/blame/74774e46375b023f73dcc9085339aa56ae7a9675/internal/download/downloader.go#L65-L73

In the meanwhile, if you can somehow flatten your structure or use TAR ––we don't need Windows archives to be ZIP necessarily, you can move forward.

vharsh commented:
> In the meanwhile, if you can somehow flatten your structure or use TAR ––we don't need Windows archives to be ZIP necessarily, you can move forward.

Cool, I'll flatten the release binaries and releasing a patch release `0.3.1`, update the SHAs, and remove the `files` key,

ahmetb commented:
:robot: _Beep beep! I’m a robot speaking on behalf of @ahmetb._ :robot:

-----

This pull request **does not** seem to be a straightforward version bump. I'll have a human review this.

_Why wasn't this detected as a plugin version bump:_

>file `plugins/openebs.yaml` is not a straightforward version bump: change on line `5` does not seem like a version bump: [`-    bin: kubectl-openebs_v0.2.0_Darwin_x86_64/kubectl-openebs`]

ahmetb commented:
:robot: _Beep beep! I’m a robot speaking on behalf of @ahmetb._ :robot:

-----

This pull request **does not** seem to be a straightforward version bump. I'll have a human review this.

_Why wasn't this detected as a plugin version bump:_

>file `plugins/openebs.yaml` is not a straightforward version bump: change on line `5` does not seem like a version bump: [`-    bin: kubectl-openebs_v0.2.0_Darwin_x86_64/kubectl-openebs`]

corneliusweig commented:
> Cool, I'll flatten the release binaries and releasing a patch release 0.3.1, update the SHAs, and remove the files key,

An alternative is to use tar for your windows archive too. Krew also supports it on windows. However, a flat archive is even better IMO.

/lgtm
/approve
